functions/profile.js

```diff
@@ -0,0 +1,107 @@
+const admin = require('./admin');
+const Promise = require('bluebird');
+const User = require("./user");
+
+// Trigger: /accounts/{userId}/githubToken
+// For legacy clients that only write githubToken (not profileNeedsUpdate)
+function enqueueUpdateLegacy(event) {
+  if (event.data.val()) {
+    return event.data.ref.parent.child('profileNeedsUpdate').set(true);
+  }
+}
+
+// Trigger: /accounts/{userId}/profileNeedsUpdate
+function enqueueUpdate(event) {
+  // Ignore profileNeedsUpdate=false
+  if (!event.data.val()) {
+    return;
+  }
+
+  // Create an entry in the queue
+  return writeQueueEntry(admin.database(), event.params.userId);
+}
+
+function writeQueueEntry(db, uid) {
+  const queueRef = db.ref(`/updateProfileQueue/${uid}`);
+  return queueRef.set({
+    createdAt: Math.floor(Date.now() / 1000),
+  });
+}
+
+// Trigger: /updateProfileQueue/{userId}
+function update(event) {
+  // Ignore deletes
+  if (!event.data.exists()) {
+    return;
+  }
+
+  const uid = event.params.userId;
+  const rootRef = admin.database().ref();
+
+  return User.findById(uid)
+    .then(user => {
+      // Fetch and save new profile data from GitHub
+      return user.updateProfile();
+    })
+    .then(() => {
+      // Atomically delete the queue entry and clear profileNeedsUpdate flag
+      const updates = {}
+      updates[`/updateProfileQueue/${uid}`] = null;
+      updates[`/accounts/${uid}/profileNeedsUpdate`] = false;
+      return rootRef.update(updates);
+    });
+}
+
+function cron() {
+  const db = admin.database();
+
+  function fixEnqueueUpdateLegacyFails() {
+    return db.ref('accounts').orderByChild('profileNeedsUpdate').equalTo(null).once('value')
+      .then(snapshot => {
+        var promises = [];
+        snapshot.forEach(data => {
+          if (data.val().githubToken) {
+            promises.push(db.ref(`/accounts/${data.key}/profileNeedsUpdate`).set(true));
+          }
+        });
+        return Promise.all(promises);
+      });
+  }
+
+  function fixEnqueueUpdateFails() {
+    return db.ref('accounts').orderByChild('profileNeedsUpdate').equalTo(true).once('value')
+      .then(snapshot => {
+        var promises = [];
+        snapshot.forEach(data => {
+          promises.push(writeQueueEntry(db, data.key));
+        });
+        return Promise.all(promises);
+      });
+  }
+
+  function fixUpdateFails() {
+    const staleSeconds = 60*2;
+    const staleTime = Math.floor((Date.now() / 1000) - staleSeconds);
+
+    return db.ref('updateProfileQueue').orderByChild('createdAt').endAt(staleTime).once('value')
+      .then(snapshot => {
+        const now = Math.floor(Date.now() / 1000);
+        var promises = [];
+        snapshot.forEach(data => {
+          promises.push(db.ref(`/updateProfileQueue/${data.key}/updatedAt`).set(now));
```

Comment:
local `var userId = data.key` would make this a little more clear
also `snapshot` could be more descriptive, like `staleTasks` or `staleUpdates` or something like that